proto/serialization.go

```diff
@@ -137,8 +152,14 @@ func (d *decoder) DecodeString() string {
 	return string(b)
 }
 
-func (d *decoder) DecodeArrayLen() int {
-	return int(d.DecodeInt32())
+func (d *decoder) DecodeArrayLen() (int, error) {
+	len := int(d.DecodeInt32())
+
+	if len < 0 || len > maxParseBufSize {
+		return 0, ErrInvalidArrayLen
```

Comment:
I would argue that we need a test against maxParseBufSize here.  It looks unnatural, because we compare values with two different units - length is in pieces and bufsize in bytes.   So even if check passed and len <= maxParseBufSize  the size of array we create (len * size of structure)  might be much bigger then maxParseBufSize. 

The idea of checking sizes in order to protect against malicious messages is great, but wouldn't be better to use for this purpose the first field in the header which we ignore? 
```
	// total message size
	_ = dec.DecodeInt32()

```

Reply:
Checking of the total message size only is not sufficient. Insane array length fields will lead to slices of several GB to be allocated which typically leads to instant OOM killing or killing of the container due to memory constraints. All memory allocations based on message data must be protected or the parser can be killed easily with a malicious packet. Other Kafka parsers implement similar limits.

Adding a `maxParseArrayLen` is clearly an option although the value will likely be the same.

Reply:
The insane array length is something that is happening frequently. Introduction of new versions of requests causes fields to shift and arbitrary data is interpreted as array length.

Reply:
Thanks for explanation